apps/settings/js/modules/base/task_scheduler.js

```diff
@@ -0,0 +1,139 @@
+  
+  /**
+   * @class TaskScheduler
+   * @returns {TaskScheduler}
+   */
+  var TaskScheduler = Module.create(function() {
+    this.super(Observable).call(this);
+    this.isLocked = false;
+    this._tasks = [];
+  }).extend(Observable);
+
+  Observable.defineObservableProperty(TaskScheduler.prototype, 'isLocked', {
+    value: false
+  });
+
+  /**
+   * We will keep some constants here as a static variable
+   *
+   * @access public
+   * @memberOf TaskScheduler
+   */
+  TaskScheduler.TASK_TYPE = {
+    GENERAL: 0
+  };
+
+  /**
+   * Change the internal state to make sure we are locked.
+   *
+   * @access private
+   * @memberOf TaskScheduler
+   */
+  TaskScheduler.prototype._lock = function() {
+    this.isLocked = true;
+  };
+
+  /**
+   * Change the internal state to unlocked and execute next task.
+   *
+   * @access private
+   * @memberOf TaskScheduler
+   */
+  TaskScheduler.prototype._unlock = function() {
+    this.isLocked = false;
+    this._executeNextTask();
+  };
+
+  /**
+   * We will remove any redundant same-type tasks and keep the remaining ones
+   * including the others which are not cancelable (general use task).
+   *
+   * @access private
+   * @memberOf TaskScheduler
+   * @param {String} type
+   * @return {Array} array of tasks
+   */
+  TaskScheduler.prototype._removeRedundantTasks = function(type) {
+    return this._tasks.filter((task) => {
+      if (!task.cancelable) {
+        return true;
+      } else {
+        return task.type !== type;
+      }
+    });
+  };
+
+  /**
+   * Execute the next task when unlocked.
+   *
+   * @access private
+   * @returns {Promise}
+   * @memberOf TaskScheduler
+   */
+  TaskScheduler.prototype._executeNextTask = function() {
+    if (this.isLocked) {
+      return;
+    }
+    var nextTask = this._tasks.shift();
+    if (nextTask) {
+      this._lock();
+      return nextTask.func().then((result) => {
+        nextTask.defer.resolve(result);
+        this._unlock();
+      }, () => {
+        nextTask.defer.reject();
+        this._unlock();
+      });
+    }
+  };
+
+  /**
+   * This is the only entry point for caller, and we will enqueue all tasks
+   * one by one and execute them in order.
+   *
+   * @access public
+   * @memberOf TaskScheduler
+   * @param {Object} task
+   * @param {String} task.type
+   * @param {Function} task.func - make sure the func would return a promise
+   * @param {Boolean} task.cancelable
+   * @return {Promise}
+   */
+  TaskScheduler.prototype.enqueue = function(task) {
+    var defer = Defer();
+
+    // if there is no func, let's directly resolve this task
+    if (!task.func) {
+      return Promise.resolve();
+    }
+
+    if (!task.type) {
+      task.type = TaskScheduler.TASK_TYPE.GENERAL;
+    }
+
+    task.defer = defer;
+    this._tasks = this._removeRedundantTasks(task.type);
+    this._tasks.push(task);
+    this._executeNextTask();
+
+    return defer.promise;
+  };
+
+  return function ctor_task_scheduler() {
```

Comment:
The function returned from `Module.create` is a normal function that returns an object when being called. We can simply return the function.